turbo/txpool-provider/block_diff.go

```diff
@@ -0,0 +1,187 @@
+// old fork, but not on the canonical chain and ii) a list of transactions that
+// were included in both the old fork and the canonical chain.
+func cmpTxsAcrossFork(oldHead, newHead *types.Header, db ethdb.Database) (types.Transactions, types.Transactions) {
+	var discarded, included types.Transactions
+
+	if oldHead != nil {
+		getter := adapter.NewBlockGetter(db)
+		oldNum := oldHead.Number.Uint64()
+		newNum := newHead.Number.Uint64()
+
+		// If the reorg is too deep, avoid doing it.
+		if depth := uint64(math.Abs(float64(oldNum) - float64(newNum))); depth <= 128 {
+			rem := getter.GetBlock(oldHead.Hash(), oldHead.Number.Uint64())
+			add := getter.GetBlock(newHead.Hash(), newHead.Number.Uint64())
+
+			if rem == nil {
+				// This can happen if a setHead is performed, where we simply discard the old
+				// head from the chain. If that is the case, we don't have the lost
+				// transactions any more, and there's nothing to add.
+				return types.Transactions{}, types.Transactions{}
+			}
+			for rem.NumberU64() > add.NumberU64() {
+				discarded = append(discarded, rem.Transactions()...)
+				if rem = getter.GetBlock(rem.ParentHash(), rem.NumberU64()-1); rem == nil {
+					return types.Transactions{}, types.Transactions{}
+				}
+			}
+			for add.NumberU64() > rem.NumberU64() {
+				included = append(included, add.Transactions()...)
+				if add = getter.GetBlock(add.ParentHash(), add.NumberU64()-1); add == nil {
+					return types.Transactions{}, types.Transactions{}
+				}
+			}
+			for rem.Hash() != add.Hash() {
+				discarded = append(discarded, rem.Transactions()...)
+				if rem = getter.GetBlock(rem.ParentHash(), rem.NumberU64()-1); rem == nil {
+					return types.Transactions{}, types.Transactions{}
+				}
+				included = append(included, add.Transactions()...)
+				if add = getter.GetBlock(add.ParentHash(), add.NumberU64()-1); add == nil {
+					return types.Transactions{}, types.Transactions{}
+				}
+			}
+
+		}
+	}
+
+	return included, discarded
+}
+
+// buildAccountDiff wraps the response from touchedAccounts(..) in a protobuf
+// compatible struct.
+func buildAccountDiff(txs types.Transactions, chainId *big.Int, db ethdb.Database) []*pb.AccountInfo {
+	addrs, nonces, balances := touchedAccounts(txs, chainId, db)
+	diffs := make([]*pb.AccountInfo, len(addrs))
+
+	for i, addr := range addrs {
+		diff := pb.AccountInfo{
+			Address: addr.Bytes(),
+			Nonce:   i64tob(nonces[i]),
+			Balance: balances[i].Bytes(),
+		}
+
+		diffs[i] = &diff
+	}
+
+	return diffs
+}
+
+// buildAccountDiff iterates through EOAs which originate a transaction in the
+// list of transactions. Each account's latest nonce and balance is returned.
+func touchedAccounts(txs types.Transactions, chainId *big.Int, db ethdb.Database) ([]common.Address, []uint64, []*big.Int) {
+	m := make(map[common.Address]bool)
+	addrs := []common.Address{}
+	signer := types.NewEIP155Signer(chainId)
+
+	for _, tx := range txs {
+		from, _ := types.Sender(signer, tx)
```

Comment:
This can probably be avoided by using the `TxSenders` table, as was done in the TxPool stage